apps/hash-graph/bin/cli/src/subcommand/snapshot.rs

```diff
@@ -48,64 +49,45 @@ pub async fn snapshot(args: SnapshotArgs) -> Result<(), GraphError> {
             report
         })?;
 
-    let mut store = SnapshotStore::new(
-        pool.acquire_owned()
-            .await
-            .change_context(GraphError)
-            .map_err(|report| {
-                tracing::error!(error = ?report, "Failed to acquire database connection");
-                report
-            })?,
-    );
-
-    // TODO: Use a thin wrapper around serde-json which implements `Sink` and `Stream` to avoid the
-    //       need for a separate thread.
-    let (sender, mut receiver) = futures::channel::mpsc::channel::<SnapshotEntry>(8);
-    let mut sender = sender.sink_map_err(Report::new);
+    let mut store = SnapshotStore::new(pool.acquire().await.change_context(GraphError).map_err(
+        |report| {
+            tracing::error!(error = ?report, "Failed to acquire database connection");
+            report
+        },
+    )?);
 
     match args.command {
         SnapshotCommand::Dump(_) => {
-            let reader = tokio::spawn(async move { store.dump_snapshot(sender).await });
-
-            let mut stdout = BufWriter::new(io::stdout());
-            while let Some(entry) = receiver.next().await {
-                serde_json::to_writer(&mut stdout, &entry)
-                    .into_report()
-                    .change_context(GraphError)?;
-                stdout
-                    .write(&[b'\n'])
-                    .into_report()
-                    .change_context(GraphError)?;
-            }
-
-            reader
+            store
+                .dump_snapshot(
+                    FramedWrite::new(
+                        io::BufWriter::new(io::stdout()),
+                        codec::JsonLinesEncoder::default(),
+                    )
+                    .sink_map_err(|report| {
+                        report.attach_printable("Failed to write snapshot to stdout")
+                    }),
+                )
                 .await
-                .into_report()
-                .change_context(GraphError)?
-                .change_context(GraphError)?;
+                .change_context(GraphError)
+                .attach_printable("Failed to read snapshot from store")?;
```

Comment:
Is this error correct?

Reply:
I think it'd be pretty strange to see:
> Failed to write snapshot to stdout
> Failed to read snapshot from store

Next to each other. "Failed to write snapshot to stdout" sounds like there was a problem actually writing to stdout

Reply:
These two errors apply to different locations. `"Failed to write snapshot to stdout"` will be printed if `FramedWrite` is failing. `"Failed to read snapshot from store"` is printed when the `dump_snapshot` command is failing, which could be anything like a database error.

Reply:
So what happens if `FramedWrite` fails? Does it not propagate up to `dump_snapshot` failing?

Reply:
It is, but it's at a different context where it makes sense:
```
Error: the Graph query layer encountered an error during execution
├╴at bin/cli/src/subcommand/snapshot.rs:72:18
├╴Failed to read snapshot from store
│
├─▶ could not write a snapshot record into the sink
│   ╰╴at /Users/ghost/Development/hash/apps/hash-graph/lib/graph/src/snapshot.rs:142:52
│
╰─▶ test error which I manually inserted as I'm not smart enough to let writing to `stdout` fail
    ├╴at /Users/ghost/Development/hash/apps/hash-graph/lib/graph/src/snapshot/codec.rs:32:13
    ├╴entity: 63049908-fc50-43d4-81da-ae180e0f3b53%0004a6a9-0a9e-4fa9-874a-ad742d7dc5da
    ╰╴Failed to write snapshot to stdout
```

Reply:
I don't feel strongly about this, we can merge as is, but I still think this:
```
├╴Failed to read snapshot from store
│
├─▶ could not write a snapshot record into the sink
```

Is confusing, we aren't erroring while trying to read the snapshot, we're erroring while trying to write the snapshot

Reply:
Hmm, do you think this is more clear?
```
Error: the Graph query layer encountered an error during execution
├╴at bin/cli/src/subcommand/snapshot.rs:68:18
├╴Failed to read snapshot from store
│
├─▶ could not write a snapshot record into the sink
│   ├╴at /Users/ghost/Development/hash/apps/hash-graph/lib/graph/src/snapshot.rs:166:56
│   ╰╴Failed to write snapshot to stdout
│
╰─▶ deserializing entity failed
    ├╴at /Users/ghost/Development/hash/apps/hash-graph/lib/graph/src/snapshot/codec.rs:32:13
    ╰╴entity: 63049908-fc50-43d4-81da-ae180e0f3b53%0004a6a9-0a9e-4fa9-874a-ad742d7dc5da
```
Or should we just remove it entirely?

Reply:
I think we can improve this by just replacing `Failed to read snapshot from store` with  `Failed to produce snapshot dump`

Reply:
Or just
```
Error: the Graph query layer encountered an error during execution
├╴at bin/cli/src/subcommand/snapshot.rs:72:18
├╴Failed to produce snapshot dump
│
├─▶ could not write a snapshot record into the sink
│   ╰╴at /Users/ghost/Development/hash/apps/hash-graph/lib/graph/src/snapshot.rs:166:56
│
╰─▶ deserializing entity failed
    ├╴at /Users/ghost/Development/hash/apps/hash-graph/lib/graph/src/snapshot/codec.rs:32:13
    ╰╴entity: 63049908-fc50-43d4-81da-ae180e0f3b53%0004a6a9-0a9e-4fa9-874a-ad742d7dc5da
```

Reply:
Yep

Reply:
the first or the latter? 😂

Reply:
I don't feel strongly about the `Failed to write snapshot to stdout` bit. If it's possible for that (as in the stdout bit) to error, and if we think the errors won't encapsulate the information, let's add the message. Otherwise we can drop it

Reply:
I removed it. If we ever encounter errors here and the information is not helpful we can add more attachments in the appropriated places. E.g. when reading from a file we may want to attach other information. If writing to `stdout` does not work I'm sure we have some error message which is helpful. If we, however, end up with errors like `"Resource temporarily unavailable"` I don't see it as very helpful (I never encountered those issues, yet).